force-app/main/default/classes/DMLOperations.cls

```diff
@@ -144,11 +161,40 @@ public with sharing class DMLOperations {
      * @param oppNames The list of Opportunity names.
      */
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
-        // Loop through each Opportunity name
-            // Create an Opportunity with the given name
-            // Add the Opportunity to the list
-    }
-    
+        Account oppAccount;
+        List<ACCOUNT> accounts = [SELECT Id FROM ACCOUNT WHERE Name = :accountName LIMIT 1];
+        if (Accounts.isEmpty()) {
+            Account newAccount = new Account();
+            newAccount.Name = accountName;
+            Insert newAccount;
+            oppAccount = newAccount;
+        } else {
+             oppAccount = Accounts[0];
+        }
+        Map<String, Opportunity> existingOppsMap = new Map<String, Opportunity>();
+        List<Opportunity> existingOppsList = [SELECT Id, Name FROM Opportunity WHERE Name IN :oppNames];
+        for(Opportunity currOpp : existingOppsList) {
+            existingOppsMap.put(currOpp.Name, currOpp);
+        }
+        List<Opportunity> oppstoUpsert = new List<Opportunity>();
+        for (String currName : oppNames) {
+            Opportunity singleOpp;
+            if (existingOppsMap.containsKey(currName)) {
+                singleOpp = existingOppsMap.get(currName);
+                SingleOpp.AccountId = oppAccount.Id;
+            } else {
+                singleOpp = new Opportunity();
```

Comment:
Great job implementing the logic to handle a pre-existing Account and/or Opportunity records